src/huggingface_hub/inference/_common.py

```diff
@@ -259,6 +256,47 @@ def _bytes_to_image(content: bytes) -> "Image":
     return Image.open(io.BytesIO(content))
 
 
+## PAYLOAD UTILS
+
+
+def _prepare_payload(
+    inputs: Union[str, Dict[str, Any], ContentT],
+    parameters: Optional[Dict[str, Any]],
+    expect_binary: bool = False,
+) -> Dict[str, Any]:
+    """
+    Used in `InferenceClient` and `AsyncInferenceClient` to prepare the payload for an API request, handling various input types and parameters.
+    `expect_binary` is set to `True` when the inputs are a binary object or a local path or URL. This is the case for image and audio inputs.
+    """
+    if parameters is None:
+        parameters = {}
+    parameters = {k: v for k, v in parameters.items() if v is not None}
+    has_parameters = len(parameters) > 0
+
+    is_binary = isinstance(inputs, (bytes, Path))
+    # If expect_binary is True, inputs must be a binary object or a local path or a URL.
+    if expect_binary and not is_binary and not isinstance(inputs, str):
+        raise ValueError(f"Expected binary inputs or a local path or a URL. Got {inputs}")  # type: ignore
+    # Send inputs as raw content when no parameters are provided
+    if expect_binary and not has_parameters:
+        return {"data": inputs}
+    # If expect_binary is False, inputs must not be a binary object.
+    if not expect_binary and is_binary:
+        raise ValueError(f"Unexpected binary inputs. Got {inputs}")  # type: ignore
+
+    json: Dict[str, Any] = {}
+    # If inputs is a bytes-like object, encode it to base64
+    if isinstance(inputs, (bytes, Path)) or (isinstance(inputs, str) and inputs.startswith(("http://", "https://"))):
+        json["inputs"] = _b64_encode(inputs)
```

Comment:
```suggestion
    if expect_binary:
        json["inputs"] = _b64_encode(inputs)
```

Given the conditions above, we know inputs is binary or url/path if we reached this point with `expect_binary` been True. 

The problem with `if ... or (isinstance(inputs, str) and inputs.startswith(("http://", "https://"))):` is that if a user passes a URL for a non-binary task (like text classification) we don't want to fetch that URL.

Reply:
yes my bad! I thought I pushed this fix, I git stash these changes when I switched to something else and forgot it 🤦‍♀️

Reply:
fixed in [9ac79df](https://github.com/huggingface/huggingface_hub/pull/2601/commits/9ac79df56507586cfada4853afcd33a63dc872f7).